### PR TITLE
Only backup bios region when only writing the bios region

### PR DESCRIPTION
**Pull request by Sean-StarLabs**

### CODE REVIEW

**Comment by Sean-StarLabs** [comment]
This works and makes it build with the latest flashrom for focal and hirsute.

Before I tidy, I'd guess that @superm1 won't want the dependency changes?

**Comment by superm1** [comment]
### Thanks for your commit
The fwupd project has had a style change starting with fwupd 1.6.x.  The style is summarized in https://github.com/fwupd/fwupd/blob/master/CONTRIBUTING.md and it appears your current pull request is inconsistent with the new style.  However to keep bisect useful code in the project is only changed to the new style when pull requests are opened.

**Clang Format**
Fortunately we have added a `.clang-format` and various helper scripts to help you fix your pull request.  Follow these steps to configure the environment and fix your code.
1. Run `./contrib/setup` to set up [pre-commit](https://pre-commit.com/) locally.  This will help ensure that you have all the build dependencies necessary and automatically reformat code in the future.
2. Run `./contrib/reformat-code.py master` to reformat all code that has changed in your local branch relative to master
3. Run `git commit add .; git commit --amend` or use `git rebase -i` to amend your commit(s).
 
**Pre-commit overview**
The pre-commit rules are available at https://github.com/fwupd/fwupd/blob/master/.pre-commit-config.yaml and will check all languages and text for style.  After you have run `./contrib/setup` any future commits will check the diff relative to all of these rules.

**Comment by superm1** [comment]
> This works and makes it build with the latest flashrom for focal and hirsute.
> 
> Before I tidy, 
I would suggest following the other comment and running `contrib/setup`.  

> I'd guess that @superm1 won't want the dependency changes?

If they make sense they should absolutely be in a separate commit at least.  They seem unrelated.

File: plugins/flashrom/fu-flashrom-internal-device.c [hunk 1]
```diff
@@ -52,6 +52,7 @@ fu_flashrom_internal_device_prepare (FuDevice *device,
 		FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
 		struct flashrom_flashctx *flashctx = fu_flashrom_device_get_flashctx (parent);
 		gsize flash_size = fu_flashrom_device_get_flash_size (parent);
+		struct flashrom_layout *layout;
```
Comment (superm1) [inline plugins/flashrom/fu-flashrom-internal-device.c:55]: did you forget to push more?

Comment (Sean-StarLabs) [reply plugins/flashrom/fu-flashrom-internal-device.c:55]: No, BtG + IFWI used all the brain I had available today, will do it tomorrow!

Comment (hughsie) [reply plugins/flashrom/fu-flashrom-internal-device.c:52]: Maybe it makes sense to set the pointer to NULL initially?

File: plugins/flashrom/fu-flashrom-internal-device.c [hunk 2]
```diff
@@ -49,10 +49,31 @@ fu_flashrom_internal_device_prepare(FuDevice *device, FwupdInstallFlags flags, G
...
+		if (flashrom_layout_read_from_ifd(&layout, flashctx, NULL, 0)) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_READ,
+					    "failed to read layout from Intel ICH descriptor");
+			return FALSE;
+		}
+
+		/* include bios region for safety reasons */
+		if (flashrom_layout_include_region(layout, "bios")) {
```
[the remark this answers is not in the document]
Comment (Sean-StarLabs) [reply plugins/flashrom/fu-flashrom-internal-device.c:66]: Hmm, I've only been thinking about recent scenarios.

Thinking out loud, Flashrom can program:
* BIOS
* ME (if it's disabled)
* descriptor (depend on ROM config),
* EC (at the bottom of my to-do list - save using flashrom and superio separately and just have one update
And a tonne of other things.

coreboot can be built with or without ME, Descriptor and EC regions but will always have a BIOS region.

I think we're the only ones using this atm, but maybe to cover more use cases:

If there's a BIOS region, we have to flash that. If it fails, errors...
In this case, try and flash all other regions. If they fail, that's fine but show a warning.

If there isn't a BIOS region, flash all regions but if any fail, errors.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/flashrom/fu-flashrom-internal-device.c:66]: I think I'd feel a lot more _warm and fuzzy_ if this was behind a flag in the .quirk file controlled with `fu_device_register_private_flag`....

[the remark this answers is not in the document]
Comment (Sean-StarLabs) [reply plugins/flashrom/fu-flashrom-internal-device.c:66]: Gotcha. Defacto region to be the BIOS though?

At the moment, it's only ever writing to the BIOS region - all the PR is doing is making it only backup the BIOS region.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/flashrom/fu-flashrom-internal-device.c:66]: > all the PR is doing is making it only backup the BIOS region

Ahh! That bit didn't compute. I guess this patch is safe in all cases then? If so, can you fixup the commit mesage to be a bit more clear, rebase against master and then force push to update this PR pls.

[the remark this answers is not in the document]
Comment (Sean-StarLabs) [reply plugins/flashrom/fu-flashrom-internal-device.c:66]: Yup, just a bug fix.

**Review by hughsie — APPROVED** [review]

**Pull request merged**